### ISSUE TITLE
Trouble proxying requests

### ISSUE THREAD

**Issue by brandonparsons** [ember-cli/ember-cli#2871]
Hi there,

Starting up a new ember project (after a few months away) and having some issues proxying requests via ember-cli.  This could very well be something I am doing wrong.....

I have a pretty plain vanilla ember JS app, and a rails app running on `localhost:3000`.  I have verified that I can visit `localhost:3000/api/v1/objects` in both the command line & the browser, both properly return JSON.

The ember app is started as follows: `ember serve --proxy http://localhost:3000`, and I get an error the second it tries to proxy a request:

```
$ ember serve --proxy http://127.0.0.1:3000
version: 0.1.5
Proxying to http://127.0.0.1:3000
Livereload server on port 35729
Serving on http://0.0.0.0:4200/

Build successful - 1877ms.

Slowest Trees                  | Total          
-------------------------------+----------------
EsnextFilter                   | 375ms          
Class                          | 300ms          
Class                          | 206ms          
ES3SafeFilter                  | 146ms          
JSHint - App                   | 120ms          

Error proxying to http://127.0.0.1:3000
connect ECONNREFUSED
Error: connect ECONNREFUSED
    at errnoException (net.js:904:11)
    at Object.afterConnect [as oncomplete] (net.js:895:19)
```

Is there something I need to add, or something that has changed in the way requests are supposed to be proxied?  There isn't much in the ember app itself other than a model, and some basic routes/templates.

**Comment by jakecraige** [comment]
Are you sure your computer responds to 127.0.0.1 properly? Try going to [http://127.0.0.1:3000/api/v1/objects](http://127.0.0.1:3000/api/v1/objects) and make sure that works first

**Comment by brandonparsons** [comment]
Hi @jakecraige - thanks for the suggestion. I had `127.0.0.1` there just because I had tried that out as well. I get the same error with `localhost`, and can confirm that the rails server properly responds.

```
$ ember serve --proxy http://localhost:3000
version: 0.1.5
Proxying to http://localhost:3000
Livereload server on port 35729
Serving on http://0.0.0.0:4200/

Build successful - 1968ms.

Slowest Trees                  | Total          
-------------------------------+----------------
EsnextFilter                   | 404ms          
Class                          | 320ms          
Class                          | 177ms          
ES3SafeFilter                  | 172ms          
JSHint - App                   | 136ms          

Error proxying to http://localhost:3000
connect ECONNREFUSED
Error: connect ECONNREFUSED
    at errnoException (net.js:904:11)
    at Object.afterConnect [as oncomplete] (net.js:895:19)
```

**Comment by stefanpenner** [comment]
what happens when you curl it.

```
curl  http://localhost:3000
```

**Comment by brandonparsons** [comment]
@stefanpenner - I get the HTML output of my rails' app home page.

**Comment by brandonparsons** [comment]
The API section of the rails app is under `/api/v1

**Comment by stefanpenner** [comment]
I can with confidence say this does work as I am using this feature as we speak.

I suspect something small , maybe config related is trolling

**Comment by brandonparsons** [comment]
Haha I suspected as much, I have used this before and figured it is a very widely used feature.

I have a brand new `0.1.5` ember-cli application, and made the following changes:

`adapters/application.js`

```
+import DS from 'ember-data';
+
+export default DS.RESTAdapter.extend({
+  namespace: "api/v1"
+});
```

`models/post.js`

```
+import DS from 'ember-data';
+
+export default DS.Model.extend({
+  slug:                   DS.attr('string'),
+  title:                   DS.attr('string'),
+  body:               DS.attr('number'),
+});
```

`router.js`

```
 Router.map(function() {
+  this.resource('posts', function() {
+    this.route('show', { path: '/:post_id' });
+  });
 });
```

`routes/posts.js`

```
+import Ember from 'ember';
+
+export default Ember.Route.extend({
+  model: function() {
+    return this.store.find('post');
+  }
+});
```

`routes/posts/show.js`

```
+import Ember from 'ember';
+
+export default Ember.Route.extend({
+  model: function(params) {
+    return this.store.find('post', params.post_id);
+  }
+});
```

`package.json`

```
-    "ember-cli-content-security-policy": "0.3.0",
```

And I get the same error:

```
$ ember serve --proxy http://localhost:3000
version: 0.1.5
Proxying to http://localhost:3000
Livereload server on port 35729
Serving on http://0.0.0.0:4200/

Build successful - 1273ms.

Slowest Trees                  | Total          
-------------------------------+----------------
Class                          | 198ms          
EsnextFilter                   | 191ms          
EsnextFilter                   | 71ms           
Class                          | 68ms           
JSHint - App                   | 64ms           

Error proxying to http://localhost:3000
connect ECONNREFUSED
Error: connect ECONNREFUSED
    at errnoException (net.js:904:11)
    at Object.afterConnect [as oncomplete] (net.js:895:19)
Error proxying to http://localhost:3000
```

```
$ ember -v
version: 0.1.5
node: 0.10.28
npm: 2.1.8
```

**Comment by brandonparsons** [comment]
I have also tried updating node, to no avail:

```
$ ember -v
version: 0.1.5
node: 0.10.35
npm: 2.1.8
```

**Comment by brandonparsons** [comment]
I'm not even going to try to figure out why, but API proxying to a rails app works when you use Thin as an application server, but not Puma (development mode).

**Comment by indirect** [comment]
I'm also experiencing this, suddenly, on OS X 10.10, on an established ember-cli project where proxying has been working fine (with Puma) for months.

The problem isn't a difference between Thin and Puma, it's a problem with the Rails server listener bindings: Puma defaults to binding to `0.0.0.0`, while ember-cli is trying to proxy to `localhost`. In the past, those have been the same thing, but on OS X they might not be. `localhost` resolves to the IPv6 address of the machine, which is `::1`. For reasons I don't fully remember right now, binding to `0.0.0.0` doesn't bind to `::1`, and so the connection fails.

tl;dr: tell your rails server to bind to localhost with `bin/rails server --binding localhost`

**Comment by dmarkow** [comment]
@indirect Explicitly using `127.0.0.1` everywhere works, but it presents a problem if you need to connect on the private ipv4 address of your machine (e.g. if you're debugging an iOS app against a development rails server)

In my case, binding the rails server to `0.0.0.0` and leaving the ember proxy pointing to `localhost` works (and my private IP address still works too)

I think a somewhat recent change to Puma did this, because I also had this working perfectly fine for many months on OS X 10.10/puma. Switching to thin (or webrick) works fine.

**Comment by brandonparsons** [comment]
Yeah I am just using Thin in development, and deploying with Puma. Obviously not ideal in terms of dev/prod parity, but you aren't running with multiple threads in development anyway so you are already kinda borked from that perspective :)

**Comment by davekaro** [comment]
I ran into this same problem and everything described is exactly my experience. I wanted to note one thing I noticed that nobody else mentioned. When booting rails with puma, the output says `Rails 4.2.0 application starting in development on http://localhost:3000` and then later it says `* Listening on tcp://localhost:3000`. With webrick or thin, they don't have the message about listening on a `tcp` URI. Thin will say `Listening on localhost:3000, CTRL+C to stop` and webrick doesn't have a listen message.

It is odd though because visiting the rails app in the browser still works fine. But I suspect the way Puma "binds to tcp://" is causing problems with the way ember-cli attempts to connect.

**Comment by joshuairl** [comment]
I'm getting this now suddenly.
404 not found.

Rails log doesn't even see the request so it's obviously not even proxying to the rails server.

**Comment by joshuairl** [comment]
OMG ignore me, long day... --proxy http:/localhost:4300 definitely not going to work with one `/`

**Comment by erickrause** [comment]
This doesn't look to have been resolved.  But Rails 4.2 recently changed how it would respond to requests locally.  See: http://edgeguides.rubyonrails.org/4_2_release_notes.html#default-host-for-rails-server  
I started my rails server with the following and it worked: rails server -b 0.0.0.0

**Comment by rmariuzzo** [comment]
I'm trying to proxy all XHR request to a Laravel 5 app. Using the following command didn't work (got ECONNREFUSED) in Mac OS X 10.10.3:

```shell
ember serve --proxy http://localhost:8000/
```

I solved my issue using the IPv6:

```shell
ember serve --proxy http://\[::1\]:8000/
```

**Comment by stefanpenner** [comment]
> I'm trying to proxy all XHR request to a Laravel 5 app. Using the following command didn't work (got ECONNREFUSED) in Mac OS X 10.10.3:

any idea what Laravel 5 binds to?

**Comment by robclancy** [comment]
Just had this issue with Laravel 4.x (would work the same as 5) and all I had to do to fix it was to run laravel with `127.0.0.1` instead of letting it run with the default `localhost`.

`./artisan serve --host 127.0.0.1`

**Comment by rchampourlier** [comment]
I'm using Grape as a backend, mounted as a Rack endpoint through a `config.ru` setup. I've tried several configs:

1. **Rack server using Webrick on port 9292 ran through `rackup`**
Any attempt to access the data using `ember server --proxy localhost:9292`, `ember server --proxy 127.0.0.1:9292` or event `ember server --proxy http://\[::1\]:9292` would fail with a 404, even if the Rack server correctly logs the call with the same signature as a `curl localhost:9292`... that works as expected.

2. **Puma on port 5000 (default) ran through `foreman start` with the following `Procfile`**
`web: bundle exec puma -C config/puma.rb`

and `ember server --proxy http://127.0.0.1:5000` works!

It's a shame I can't get to understand why... and it makes me lose the auto-reload of the server code.

**Comment by rchampourlier** [comment]
GOT IT!

I'm using [guard-shotgun](https://github.com/rchampourlier/guard-shotgun) to run my Rack server with auto-reload. I specified some params so that the Rack server is ran explicitely with `127.0.0.1` as host and the `puma` server instead of Webrick. In this server configuration, `ember server --proxy http://127.0.0.1:3000` will work, no more 404. The same explicit `127.0.0.1` host with Webrick however will not work...

**Comment by constantm** [comment]
Just ran into the exact same issue while using Puma. Using the IPV6 address as advised by @rmariuzzo works perfectly. So bizarre.

**Comment by ekampp** [comment]
I found that it's a problem with rails not listening to the 0.0.0.0 and 127.0.0.1 interfaces. Adding the `-b 0.0.0.0` flag to the `rails s` command solved the problem for me.
